src/raw_mem.rs

```diff
@@ -0,0 +1,282 @@
+    }
+}
+
+pub trait RawMem {
+    type Item;
+
+    fn allocated(&self) -> &[Self::Item];
+    fn allocated_mut(&mut self) -> &mut [Self::Item];
+
+    /// # Safety
+    /// Caller must guarantee that `fill` makes the uninitialized part valid for
+    /// [`MaybeUninit::slice_assume_init_mut`]
+    ///
+    /// ### Incorrect usage
+    /// ```no_run
+    /// # #![feature(allocator_api)]
+    /// # use std::alloc::Global;
+    /// # use std::mem::MaybeUninit;
+    /// # use platform_mem::Result;
+    /// use platform_mem::{Alloc, RawMem};
+    ///
+    /// let mut alloc = Alloc::new(Global);
+    /// unsafe {
+    ///     alloc.grow(10, |_uninit: &mut [MaybeUninit<u64>]| {
+    ///         // `RawMem` relies on the fact that we initialize memory
+    ///         // even if they are primitives
+    ///     })?;
+    /// }
+    /// # Result::Ok(())
+    /// ```
+    unsafe fn grow(
+        &mut self,
+        cap: usize,
+        fill: impl FnOnce(&mut [MaybeUninit<Self::Item>]),
+    ) -> Result<&mut [Self::Item]>;
+
+    /// [`grow`] which assumes that the memory is already initialized
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that one of the following is true:
+    ///
+    /// * memory already initialized as [`Item`]
+    ///
+    /// * memory is initialized bytes and [`Item`] can be represented as bytes
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use platform_mem::Result;
+    /// use platform_mem::{FileMapped, RawMem};
+    ///
+    /// let mut file = FileMapped::from_path("..")?;
+    /// // file is always represents as initialized bytes
+    /// // and usize is transparent as bytes
+    /// let _: &mut [usize] = unsafe { file.grow_assumed(10)? };
+    /// # Result::Ok(())
+    /// ```
+    ///
+    /// [`grow`]: Self::grow
+    /// [`Item`]: Self::Item
+    unsafe fn grow_assumed(&mut self, cap: usize) -> Result<&mut [Self::Item]> {
+        self.grow(cap, |_| {})
+    }
+
+    /// # Safety
+    /// [`Item`] must satisfy [initialization invariant][inv] for [`mem::zeroed`]
+    ///
+    /// [`Item`]: Self::Item
+    ///  [inv]: MaybeUninit#initialization-invariant
+    ///
+    /// # Examples
+    /// Correct usage of this function: initializing an integral-like types with zeroes:
+    /// ```
+    /// # #![feature(allocator_api)]
+    /// # use platform_mem::Error;
+    /// use platform_mem::{Global, RawMem};
+    ///
+    /// let mut alloc = Global::new();
+    /// let zeroes: &mut [(u8, u16)] = unsafe {
+    ///     alloc.grow_zeroed(10)?
+    /// };
+    ///
+    /// assert_eq!(zeroes, [(0, 0); 10]);
+    /// # Ok::<_, Error>(())
+    /// ```
+    ///
+    /// Incorrect usage of this function: initializing a reference with zero:
+    /// ```no_run
+    /// # #![feature(allocator_api)]
+    ///  # use platform_mem::Error;
+    /// use platform_mem::{Global, RawMem};
+    ///
+    /// let mut alloc = Global::new();
+    /// let zeroes: &mut [&'static str] = unsafe {
```

Comment:
Remove useless name binding. It is also `UB`
```suggestion
    /// let _: &mut [&'static str] = unsafe {
```